src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs

```diff
@@ -202,7 +202,14 @@ private async Task StartHostAsync(CancellationToken cancellationToken, int attem
                     return;
                 }
 
-                var orphanTask = Orphan(localHost, logger)
+                if (isActiveHost)
+                {
+                    // We don't want to return disposed services via the Services property, so
+                    // set this to null before calling Orphan().
+                    _host = null;
```

Comment:
Currently the only consumer of Services is the ScriptHostRequestServiceProviderMiddleware, right? So previously we might be returning services that were in the process of being disposed. Now we'll be returning null. Won't that also prevent the middleware from adding its required services? If requests come in while _host is null, what happens to those requests? Are there race conditions here? What happens if _host is not yet orphaned, it's services get added to the request, then it is orphaned concurrently?

Reply:
In this scenario, the services aren't disposed while the request is in-flight; they're disposed before the request begins (or at least, before that middleware is hit). For whatever reason (single core machine, most likely), we haven't spun up a new host yet so we can sit there for some period with a disposed `_host`, which the middleware has access to. That seems bad, which this is trying to address.

A couple answers first:
- The middleware can handle null `Services`. Rather than adding the services from the JobHost, it adds the WebHost's services (which happens by default; that middleware overrides the default behavior). If some controller/middleware needs services from a JobHost won't get them in this case. Today, even requests that don't care about JobHost services (e.g.. key retrieval) will blow up here if the host is disposed. However:
  - All non-admin routes will have the `HostAvailabilityCheckMiddleware` run first, greatly reducing the chances of null `Services` on those requests. It's still there, however.
  - I suspect -- but I'm not sure how to verify -- that admin routes do not require any JobHost services, which is why we have the `HostAvailibilityCheckMiddleware` skipped in those cases. @fabiocav do you know if this assumption is correct?
- The host orphaning mid-request is handled here: https://github.com/Azure/azure-functions-host/blob/dev/src/WebJobs.Script.WebHost/DependencyInjection/ScopedResolver.cs#L30-L35. A request will have at least one child scope, and that disposal will be delayed for 5 seconds, or until the request has completed and the child scopes are disposed. I don't think we can really hold on for much longer than that, but @fabiocav added that code a while back.

All that being said -- you're right, Services will now be null more than it was previously. But the `HostAvailabilityCheckMiddleware` will protect all non-admin routes from that case (although there's a race there as well; if a host shuts down between those two middlewares). And as long as the admin routes don't require JobHost services, we should be alright here.

Reply:
Right, by design admin routes don't (can't) require a running host instance. So they only use WebHost scoped services.

I agree that the changes you've made here are good - we definitely don't want to be returning disposed services. I still worry about race conditions. I don't understand the details of the child scope disposal you referenced - perhaps that addresses these issue, it's just not clear to me.